Index: qemu/hw/usb/host-bsd.c
===================================================================
--- qemu.orig/hw/usb/host-bsd.c	2012-05-24 20:19:08.825346337 -0700
+++ qemu/hw/usb/host-bsd.c	2012-05-24 20:19:25.709430064 -0700
@@ -35,7 +35,7 @@
 
 #include <sys/ioctl.h>
 #ifndef __DragonFly__
-#include <dev/usb/usb.h>
+#include <legacy/dev/usb/usb.h>
 #else
 #include <bus/usb/usb.h>
 #endif
Index: qemu/configure
===================================================================
--- qemu.orig/configure	2012-05-24 20:19:08.825346337 -0700
+++ qemu/configure	2012-05-24 20:19:39.145496689 -0700
@@ -2305,6 +2305,14 @@
 #include <stddef.h>
 #include <fcntl.h>
 
+#if defined __stub_futimens || defined __stub___futimens
+choke me
+#endif
+
+#if defined __stub_utimensat || defined __stub___utimensat
+choke me
+#endif
+
 int main(void)
 {
     utimensat(AT_FDCWD, "foo", NULL, 0);
