Subject: Use -r instead of -Wl,-r to fix building with PIE
Author: Adrian Bunk <bunk@stusta.de>
Bug-Debian: https://bugs.debian.org/837574
Date: Tue, 25 Oct 2016 11:31:24 +0300

--- qemu-2.7+dfsg.orig/rules.mak
+++ qemu-2.7+dfsg/rules.mak
@@ -92,7 +92,7 @@ module-common.o: CFLAGS += $(DSO_OBJ_CFL
 	$(if $(findstring /,$@),$(call quiet-command,cp $@ $(subst /,-,$@), "  CP    $(subst /,-,$@)"))
 
 
-LD_REL := $(CC) -nostdlib -Wl,-r $(LD_REL_FLAGS)
+LD_REL := $(CC) -nostdlib -r $(LD_REL_FLAGS)
 
 %.mo:
 	$(call quiet-command,$(LD_REL) -o $@ $^,"  LD -r $(TARGET_DIR)$@")
