diff --git a/roms/qboot/include/bios.h b/roms/qboot/include/bios.h
index f36638b..7f8f677 100644
--- a/roms/qboot/include/bios.h
+++ b/roms/qboot/include/bios.h
@@ -1,7 +1,7 @@
 #ifndef BIOS_H_
 #define BIOS_H_
 
-#include <inttypes.h>
+#include <stdint.h>
 #include <stddef.h>
 #include <stdbool.h>
 
diff --git a/roms/qboot/malloc.c b/roms/qboot/malloc.c
index 8738373..bd0ac0f 100644
--- a/roms/qboot/malloc.c
+++ b/roms/qboot/malloc.c
@@ -1,4 +1,4 @@
-#include <inttypes.h>
+#include <stdint.h>
 #include "string.h"
 #include "bios.h"
 
