One n too many for running, need we say more.

Index: qemu/target-i386/kvm.c
===================================================================
--- qemu.orig/target-i386/kvm.c	2011-07-02 21:25:10.000000000 -0700
+++ qemu/target-i386/kvm.c	2011-07-02 21:25:20.000000000 -0700
@@ -1537,7 +1537,7 @@
                 code);
         if (host_supports_vmx() && code == VMX_INVALID_GUEST_STATE) {
             fprintf(stderr,
-                    "\nIf you're runnning a guest on an Intel machine without "
+                    "\nIf you're running a guest on an Intel machine without "
                         "unrestricted mode\n"
                     "support, the failure can be most likely due to the guest "
                         "entering an invalid\n"
