Index: qemu/hw/pc.c
===================================================================
--- qemu.orig/hw/pc.c	2011-02-18 09:56:51.000000000 -0800
+++ qemu/hw/pc.c	2011-02-18 10:03:39.000000000 -0800
@@ -536,8 +536,9 @@
         /* LGPL'ed VGA BIOS messages */
     case 0x501:
     case 0x502:
+        /* according to documentation, these can be safely ignored */
         fprintf(stderr, "VGA BIOS panic, line %d\n", val);
-        exit(1);
+        break;
     case 0x500:
     case 0x503:
 #ifdef DEBUG_BIOS
